projects/components/src/filtering/filter/parser/types/abstract-filter-parser.ts

```diff
@@ -0,0 +1,94 @@
+import { FilterOperator, fromUrlFilterOperator, toUrlFilterOperator, UrlFilterOperator } from '../../filter-operators';
+import { ParsedFilter, SplitFilter } from '../parsed-filter';
+
+export abstract class AbstractFilterParser<TValue> {
+  public abstract supportedAttributeTypes(): FilterAttributeType[];
+  public abstract supportedOperators(): FilterOperator[];
+
+  protected abstract parseValueString(
+    attribute: FilterAttribute,
+    splitFilter: SplitFilter<FilterOperator>
+  ): TValue | undefined;
+
+  protected parseKeyString(attribute: FilterAttribute, splitFilter: SplitFilter<FilterOperator>): string | undefined {
+    return attribute.displayName.toLowerCase() !== splitFilter.lhs.toLowerCase() ? undefined : attribute.name;
+  }
+
+  public parseFilterString(attribute: FilterAttribute, filterString: string): ParsedFilter<TValue> | undefined {
+    const splitFilter: SplitFilter<FilterOperator> | undefined = AbstractFilterParser.splitFilterStringByOperator(
+      this.supportedOperators(),
+      filterString,
+      true
+    );
+
+    return splitFilter !== undefined ? this.parseSplitFilter(attribute, splitFilter) : undefined;
+  }
+
+  public parseUrlFilterString(attribute: FilterAttribute, urlFilterString: string): ParsedFilter<TValue> | undefined {
+    const splitUrlFilter: SplitFilter<UrlFilterOperator> | undefined = AbstractFilterParser.splitFilterStringByOperator(
+      this.supportedOperators().map(toUrlFilterOperator),
+      decodeURIComponent(urlFilterString),
+      false
+    );
+
+    return splitUrlFilter !== undefined
+      ? this.parseSplitFilter(attribute, {
+          lhs: attribute.displayName,
+          operator: fromUrlFilterOperator(splitUrlFilter.operator),
+          rhs: splitUrlFilter.rhs
+        })
+      : undefined;
+  }
+
+  private parseSplitFilter(
+    attribute: FilterAttribute,
+    splitFilter: SplitFilter<FilterOperator>
+  ): ParsedFilter<TValue> | undefined {
+    const parsedKey = this.parseKeyString(attribute, splitFilter);
+
+    if (parsedKey === undefined) {
+      // Unable to parse attribute from lhs
+      return undefined;
+    }
+
+    const parsedValue = this.parseValueString(attribute, splitFilter);
+
+    if (parsedValue === undefined) {
+      // Unable to parse value from rhs
+      return undefined;
+    }
+
+    // Successfully parsed URL filter
+
+    return {
+      field: attribute.name,
+      operator: splitFilter.operator,
+      value: parsedValue
+    };
+  }
+
+  public static splitFilterStringByOperator<TOperator extends string>(
+    possibleOperators: TOperator[],
+    filterString: string,
+    expectSpaceAroundOperator: boolean = true
+  ): SplitFilter<TOperator> | undefined {
+    const matchingOperator = possibleOperators
+      .sort((o1: string, o2: string) => o2.length - o1.length) // Sort by length to check multichar ops first
+      .map(op => (expectSpaceAroundOperator ? ` ${op as string} ` : op))
```

Comment:
could we make this work without the flag and just collapse any potential whitespace below?

Reply:
We are using the whitespace to help us find the operator and split on it without inadvertently splitting on operators that may be embedded in the value, for instance `=` is often in url values. The user string will always have operators surrounded by spaces, so we use that as our string split token.

The exception is for URL filters, which have our underscore version of the operator, like `_eq_`, thus the reason for needing the arg flag.